test/test_torch.py

```diff
@@ -11040,6 +11040,77 @@ def test_exponential(self, device, dtype):
         with self.assertRaises(RuntimeError):
             torch.empty((1,), device=device, dtype=dtype).exponential_(-0.5)
 
+    @unittest.skipIf(not TEST_NUMPY, "Numpy not found")
+    @dtypes(*(torch.testing.get_all_fp_dtypes(include_half=False) +
+              torch.testing.get_all_complex_dtypes()))
+    @dtypesIfCUDA(*(torch.testing.get_all_fp_dtypes(include_half=True) +
+                    torch.testing.get_all_complex_dtypes()))
+    def test_exp(self, device, dtype):
+        for v in (2, -2) + ((1j, 1 + 1j) if dtype.is_complex else ()):
+            if dtype == torch.bfloat16:
+                # Currently multiply a bfloat16 type with floating-point causes error:
+                #   RuntimeError: dtype != ScalarType::Undefined INTERNAL ASSERT FAILED at
+                #   "/pytorch/aten/src/ATen/native/TensorIterator.cpp":125, please report a bug to PyTorch.
+                # We skip bfloat16 for now, but we should fix it. https://github.com/pytorch/pytorch/issues/40580
+                if self.device_type == 'cpu' or self.device_type == 'cuda':
+                    with self.assertRaises(RuntimeError):
+                        torch.tensor(v, dtype=dtype, device=device) * torch.arange(18, device=device)
+                    return
+
+            a = torch.tensor(v, dtype=dtype, device=device) * torch.arange(18, device=device) / 3 * math.pi
+            a = a.to(dtype)
+            self.compare_with_numpy(torch.exp, np.exp, a)
+
+            if dtype.is_complex:
+                inf_real_zero_imag_in = torch.tensor(complex(float('inf'), 0), device=device, dtype=dtype)
+                inf_real_zero_imag_out = torch.exp(inf_real_zero_imag_in).item()
+                self.assertTrue(math.isinf(inf_real_zero_imag_out.real))
+                if device.startswith('cpu') and not IS_WINDOWS:  # Windows tests don't show some bugs consistently
```

Comment:
I'm still not in love with how many separate cases are in this part of the test (as opposed to just declaring vals = (complex(...), complex(...), ...) and calling compare_with_numpy on the lot. But I guess this is a little more specific. If you change the `device.startswith()` calls to `self.device_type ==` and don't exclude XLA testing I guess it's OK.

Reply:
To be clear, what I was thinking was:

```
vals = (complex(...), complex...)
# see issue ...
if self.device_type == 'cpu':
  with self.assertRaises(...):
    self.compare_with_numpy(...)
else:
  self.compare_with_numpy(...)
```

You can exclude the test from running on Windows instead of adding conditionals for it, too, if Windows is flaky.
```

Reply:
If excluding Windows entirely from the test, then CUDA might not be properly tested. I have drop the `IS_WINDOWS` clause down one level, which should make the code look better.